RecoTracker/TkHitPairs/plugins/HitPairEDProducer.cc

```diff
@@ -0,0 +1,170 @@
+    throw cms::Exception("Configuration") << "HitPairEDProducer requires either produceIntermediateHitDoublets or produceSeedingHitSets to be True. If neither are needed, just remove this module from your sequence/path as it doesn't do anything useful";
+
+  if(produceSeedingHitSets_)
+    produces<RegionsSeedingHitSets>();
+  if(produceIntermediateHitDoublets_)
+    produces<IntermediateHitDoublets>();
+}
+
+void HitPairEDProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("seedingLayers", edm::InputTag("seedingLayersEDProducer"));
+  desc.add<edm::InputTag>("trackingRegions", edm::InputTag("globalTrackingRegionFromBeamSpot"));
+  desc.add<edm::InputTag>("clusterCheck", edm::InputTag("clusterCheckerEDProducer"));
+  desc.add<bool>("produceSeedingHitSets", false);
+  desc.add<bool>("produceIntermediateHitDoublets", false);
+  desc.add<unsigned int>("maxElement", 0); // default is really 0? Also when used from CombinedHitTripletGenerator?
+
+  descriptions.add("hitPairEDProducer", desc);
+}
+
+void HitPairEDProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<bool> hclusterCheck;
+  iEvent.getByToken(clusterCheckToken_, hclusterCheck);
+  const bool clusterCheckOk = *hclusterCheck;
+
+  edm::Handle<SeedingLayerSetsHits> hlayers;
+  iEvent.getByToken(seedingLayerToken_, hlayers);
+  const auto& layers = *hlayers;
+  if(layers.numberOfLayersInSet() < 2)
+    throw cms::Exception("Configuration") << "HitPairEDProducer expects SeedingLayerSetsHits::numberOfLayersInSet() to be >= 2, got " << layers.numberOfLayersInSet();
+
+  edm::Handle<edm::OwnVector<TrackingRegion> > hregions;
+  iEvent.getByToken(regionToken_, hregions);
+  const auto& regions = *hregions;
+
+  std::unique_ptr<RegionsSeedingHitSets> seedingHitSets;
+  std::unique_ptr<IntermediateHitDoublets> intermediateHitDoublets;
+  if(produceSeedingHitSets_)
+    seedingHitSets = std::make_unique<RegionsSeedingHitSets>();
+  if(produceIntermediateHitDoublets_)
+    intermediateHitDoublets = std::make_unique<IntermediateHitDoublets>(&layers);
+
+  if(!clusterCheckOk) {
+    if(produceSeedingHitSets_)
+      iEvent.put(std::move(seedingHitSets));
+    if(produceIntermediateHitDoublets_)
+      iEvent.put(std::move(intermediateHitDoublets));
+    return;
+  }
+
+  if(produceSeedingHitSets_)
+    seedingHitSets->reserve(regions.size(), localRA_.upper());
+  if(produceIntermediateHitDoublets_)
+    intermediateHitDoublets->reserve(regions.size(), layers.size());
+
+
+  LogDebug("HitPairEDProducer") << "Creating doublets for " << regions.size() << " and " << layers.size() << " layer sets";
+
+  // This is the easiest way to extract the layer pairs from the full
+  // set of seeding layers. It feels a bit stupid to do it for each
+  // event given the input is defined in configuration. Maybe do it
+  // once-per-job in SeedingLayerSetsEDProducer?
```

Comment:
Note to self: the suggestion above does not work for CA, will need a special mechanism here that works for both triplets and CA.